File: Integrations/sdks/python/scriptbee_analysis_sdk/_internal/endpoints/context_router.py
import io

from fastapi import APIRouter, Depends
from fastapi.responses import StreamingResponse

from scriptbee_analysis_sdk._internal.endpoints.contracts.web_models import (
    WebContextGraphEdge,
    WebContextGraphNode,
    WebContextGraphResponse,
    WebContextSlice,
    WebGenerateClassesRequest,
    WebGetContextResponse,
    WebLinkContextCommand,
    WebLoadContextCommand,
)
from scriptbee_analysis_sdk.abstractions.clear_context_use_case import ClearContextUseCase
from scriptbee_analysis_sdk.abstractions.generate_classes_use_case import GenerateClassesUseCase
from scriptbee_analysis_sdk.abstractions.get_context_graph_use_case import GetContextGraphUseCase
from scriptbee_analysis_sdk.abstractions.get_context_use_case import GetContextUseCase
from scriptbee_analysis_sdk.abstractions.link_context_use_case import LinkContextUseCase
from scriptbee_analysis_sdk.abstractions.load_context_use_case import LoadContextUseCase
from scriptbee_analysis_sdk.domain.project import FileId
from scriptbee_analysis_sdk.file_bundler import FileBundler


def make_context_router(
    get_context: callable,
    load_context: callable,
    link_context: callable,
    clear_context: callable,
    get_context_graph: callable,
    generate_classes: callable,
) -> APIRouter:
    router = APIRouter(tags=["Context"])

    @router.get(
        "/api/context",
        name="Context",
        summary="Get analysis context",
        description="Retrieves the current data context from the analysis service.",
    )
    def get_context_handler(
        use_case: GetContextUseCase = Depends(get_context),
    ) -> WebGetContextResponse:
        slices = use_case.get()
        return WebGetContextResponse(data=[WebContextSlice.from_context_slice(s) for s in slices])

    @router.post(
        "/api/context/load",
        status_code=204,
        name="Load",
        summary="Load data into analysis context",
        description="Loads data from files into the current analysis context using the provided loaders.",
    )
    async def load_context_handler(
        command: WebLoadContextCommand,
        use_case: LoadContextUseCase = Depends(load_context),
    ) -> None:
        files = {k: [FileId(value=f) for f in v] for k, v in command.files_to_load.items()}
        await use_case.load(files)

    @router.post(
        "/api/context/link",
        status_code=204,
        name="Link",
        summary="Link analysis context",
        description="Links the current analysis context using the provided linkers.",
    )
    async def link_context_handler(
        command: WebLinkContextCommand,
        use_case: LinkContextUseCase = Depends(link_context),
    ) -> None:
        await use_case.link(command.linker_ids)

    @router.post(
        "/api/context/clear",
        status_code=204,
        name="Clear",
        summary="Clear analysis context",
        description="Clears all data from the current analysis context.",
    )
    def clear_context_handler(
        use_case: ClearContextUseCase = Depends(clear_context),
    ) -> None:
        use_case.clear()

    @router.get(
        "/api/context/graph-nodes",
        name="GraphNodes",
        summary="Search context nodes",
        description="Searches for context nodes based on a query string.",
        tags=["Context", "Graph"],
    )
    def search_nodes(
        use_case: GetContextGraphUseCase = Depends(get_context_graph),
        query: str = "",
        offset: int = 0,
        limit: int = 10,
    ) -> WebContextGraphResponse:
        result = use_case.search_nodes(query, offset, limit)
        return WebContextGraphResponse(
            nodes=[WebContextGraphNode.from_node(n) for n in result.nodes],
            edges=[WebContextGraphEdge.from_edge(e) for e in result.edges],
        )

    @router.get(
        "/api/context/graph-nodes/{node_id}/neighbors",
        name="Neighbors",
        summary="Get node neighbors",
        description="Retrieves the immediate neighbors and edges for a specific context node.",
        tags=["Context", "Graph"],
    )
    def get_neighbors(
        node_id: str,
        use_case: GetContextGraphUseCase = Depends(get_context_graph),
    ) -> WebContextGraphResponse:
        result = use_case.get_neighbors(node_id)
        return WebContextGraphResponse(
            nodes=[WebContextGraphNode.from_node(n) for n in result.nodes],
            edges=[WebContextGraphEdge.from_edge(e) for e in result.edges],
        )

    @router.post(
        "/api/context/generate-classes",
        name="GenerateClasses",
        summary="Generate classes for analysis context",
        description="Generates script classes based on the current data context and returns them as a stream.",
    )
    async def generate_classes_handler(
        request: WebGenerateClassesRequest,
        use_case: GenerateClassesUseCase = Depends(generate_classes),
    ) -> StreamingResponse:
        languages = request.languages or []
        files = await use_case.generate_classes(languages)
        stream = io.BytesIO()
        FileBundler().write_to_stream(files, stream)
        stream.seek(0)
        return StreamingResponse(
            stream,
            media_type="application/octet-stream",
            headers={"Content-Disposition": 'attachment; filename="classes.bin"'},
        )

    return router

File: Integrations/sdks/typescript/src/_internal/endpoints/context-router.ts
import { Hono } from "hono";
import type { ClearContextUseCase } from "../../abstractions/clear-context-use-case.js";
import type { GenerateClassesUseCase } from "../../abstractions/generate-classes-use-case.js";
import type { GetContextGraphUseCase } from "../../abstractions/get-context-graph-use-case.js";
import type { GetContextUseCase } from "../../abstractions/get-context-use-case.js";
import type { LinkContextUseCase } from "../../abstractions/link-context-use-case.js";
import type { LoadContextUseCase } from "../../abstractions/load-context-use-case.js";
import type { FileId } from "../../domain/project.js";
import { FileBundler } from "../../file-bundler.js";
import {
  WebGenerateClassesRequestSchema,
  WebLinkContextCommandSchema,
  WebLoadContextCommandSchema,
  webContextGraphEdgeFromDomain,
  webContextGraphNodeFromDomain,
  webContextSliceFromDomain,
} from "./contracts/web-models.js";

export function makeContextRouter(options: {
  getContext: () => GetContextUseCase;
  loadContext: () => LoadContextUseCase;
  linkContext: () => LinkContextUseCase;
  clearContext: () => ClearContextUseCase;
  getContextGraph: () => GetContextGraphUseCase;
  generateClasses: () => GenerateClassesUseCase;
}): Hono {
  const router = new Hono();
  const fileBundler = new FileBundler();

  router.get("/api/context", (c) => {
    const slices = options.getContext().get();
    return c.json({ data: slices.map(webContextSliceFromDomain) });
  });

  router.post("/api/context/load", async (c) => {
    const body = await c.req.json();
    const parsed = WebLoadContextCommandSchema.safeParse(body);

    if (!parsed.success) {
      return c.json({ title: "Validation Failed", detail: parsed.error.message }, 400);
    }

    const filesToLoad: Record<string, FileId[]> = {};
    for (const [loaderId, fileIds] of Object.entries(parsed.data.filesToLoad)) {
      filesToLoad[loaderId] = fileIds.map((id) => ({ value: id }));
    }

    await options.loadContext().load(filesToLoad);
    return new Response(null, { status: 204 });
  });

  router.post("/api/context/link", async (c) => {
    const body = await c.req.json();
    const parsed = WebLinkContextCommandSchema.safeParse(body);

    if (!parsed.success) {
      return c.json({ title: "Validation Failed", detail: parsed.error.message }, 400);
    }

    await options.linkContext().link(parsed.data.linkerIds);
    return new Response(null, { status: 204 });
  });

  router.post("/api/context/clear", (c) => {
    options.clearContext().clear();
    return new Response(null, { status: 204 });
  });

  router.get("/api/context/graph-nodes", (c) => {
    const query = c.req.query("query") ?? "";
    const offset = parseInt(c.req.query("offset") ?? "0", 10);
    const limit = parseInt(c.req.query("limit") ?? "10", 10);

    const result = options.getContextGraph().searchNodes(query, offset, limit);
    return c.json({
      nodes: result.nodes.map(webContextGraphNodeFromDomain),
      edges: result.edges.map(webContextGraphEdgeFromDomain),
    });
  });

  router.get("/api/context/graph-nodes/:nodeId/neighbors", (c) => {
    const nodeId = c.req.param("nodeId");
    const result = options.getContextGraph().getNeighbors(nodeId);
    return c.json({
      nodes: result.nodes.map(webContextGraphNodeFromDomain),
      edges: result.edges.map(webContextGraphEdgeFromDomain),
    });
  });

  router.post("/api/context/generate-classes", async (c) => {
    const body = await c.req.json();
    const parsed = WebGenerateClassesRequestSchema.safeParse(body);

    if (!parsed.success) {
      return c.json({ title: "Validation Failed", detail: parsed.error.message }, 400);
    }

    const files = await options.generateClasses().generateClasses(parsed.data.languages ?? []);
    const bundle = fileBundler.writeToBuffer(files);

    return new Response(bundle, {
      status: 200,
      headers: {
        "Content-Type": "application/octet-stream",
        "Content-Disposition": 'attachment; filename="classes.bin"',
      },
    });
  });

  return router;
}
